backend/app/routers/authorization.py

```diff
@@ -232,7 +230,23 @@ async def set_dataset_user_role(
             ) is not None:
                 # Update if it already exists
                 auth_db = AuthorizationDB.from_mongo(auth_q)
-                if username not in auth_db.user_ids:
+                if username in auth_db.user_ids:
+                    # Only add user entry if all the others occurrences are from associated groups
+                    group_q_list = db["groups"].find(
+                        {"_id": {"$in": auth_db.group_ids}}
+                    )
+                    group_occurrences = 0
+                    async for group_q in group_q_list:
+                        group = GroupOut.from_mongo(group_q)
+                        for u in group.users:
+                            if u.user.email == username:
+                                group_occurrences += 1
+                    if auth_db.user_ids.count(username) == group_occurrences:
+                        auth_db.user_ids.append(username)
+                        await db["authorization"].replace_one(
+                            {"_id": auth_db.id}, auth_db.to_mongo()
+                        )
```

Comment:
I'm out of loop and wondering what's this "counting user in all groups" logic is for? Could you kindly explain just a little bit? Thanks!

Reply:
Each time the dataset is shared with a user or a group containing the user, that user's email is added to the list of auth_db.user_ids. This means an email might appear multiple times: <number of groups user is in> + <one occurrence for the user themselves>.

If you share the dataset with the same user repeatedly, say as a viewer, we dont want to add multiple occurrences of their email to the list, only once. BUT the user's email might be in the list 4 times thanks to being in 4 groups, but doesn't yet have an entry for their individual self. 

So we need to figure out if all the entries in the list come from groups, in which case we can add a new entry, OR if there are only n-1 entries from groups and the last entry already exists, in which case we don't add a new entry.

Reply:
Thanks for the explanation that makes sense. I was wondering why you evaluate `auth_db.user_ids.count(username) == group_occurrences` now it makes sense.